accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -616,6 +609,442 @@ It is a warning to use a null aware operator (`?.`, `?[]`, `?..`, `??`, `??=`, o
+initialization.  In all cases, the values assigned must all have the same type
+(up to nullability), or type `Null` - otherwise the variable becomes unusable
+after the join.  More precisely, the **initialization inferred type** of a
+variable after a join point is specified as follows.
+
+At a join point in the program, if an **untyped variable** has no
+**initialization inferred type** on any path into the join, then it has no
+**initialization inferred type** after the join.
+
+Otherwise, if an **untyped variable** has an **initialization inferred type** on
+at least one path (reachable or not) into the join, then we consider two cases:
+first that it has an **initialization inferred type** on every reachable path
+into the join (in which case it does not need to be made implicitly nullable),
+or that it has at least one reachable path with no **initialization inferred
+type** (in which case it must be made implicitly nullable if it is non-`late`).
+
+At a join point in the program, if an **untyped variable** has an
+**initialization inferred type** on all of the *reachable* paths into the join,
+then the **initialization inferred type** for the variable after the join point
+is computed as follows:
+  - If the **initialization inferred type** on each path which has one
+    (reachable or not) is the same type `T` (using syntactic equality), then the
+    **initialization inferred type** after the join is `T`.
+  - Otherwise, if the **initialization inferred type** on each path which has
+    one (reachable or not) is one of `T`, `T?` or `Null` for some type `T`
+    (using syntactic equality), then the **initialization inferred type** after
+    the join is `T?`.
+  - Otherwise, no consistent type can be inferred and it is an error.
+
+In other words, if an **untyped** variable of any kind is assigned on all paths
+into the join, it must be assigned either a value of the same type or its
+nullable variant, or a value of type `Null`, on all paths; and if the variable
+is assigned a value of type `Null` on any path, then the variable is treated as
+having nullable type.
+
+**Note that the case above trivially includes the case that there are no
+reachable paths into the join.**
+
+At a join point in the program, if an **untyped variable** has no
+**initialization inferred type** on at least one *reachable* path into the join,
+then the **initialization inferred type** for the variable after the join point
+is computed as follows:
+  - If the variable is declared `late` and the **initialization inferred type**
+    on each path (reachable or not) which has one is the same type `T` (using
+    syntactic equality), then the **initialization inferred type** after the
+    join is `T`.
+  - Otherwise, if the **initialization inferred type** on each path into the
+    `join` which has one is either `T`, `T?`, or `Null` for some type `T`, the
+    **initialization inferred type** after the join is `T?`.
```

Comment:
Likewise:

```dart
var x;
if (c) x = null;
// what is x here?
```

And according to what I proposed:

```dart
var x; // Unassigned+Untyped
if (c) x = null; // Assigned+Null
// Join: Either+Null
print(x); // OK, potentially assigned some unknown nullable type, treat as Null for now.
x = 123; // OK, now Assigned+Typed(int?)
x = "str"; // Error, wrong type.
```

Reply:
Also `Null`.